drivers/arm/pl011/aarch64/pl011_console.S

```diff
@@ -13,15 +13,72 @@
  */
 #include "../../../console/aarch64/console.S"
 
+	.globl	console_pl011_init
+	.globl	console_pl011_putc
+	.globl	console_pl011_getc
+	.globl	console_pl011_flush
 
-	.globl	console_core_init
-	.globl	console_core_putc
-	.globl	console_core_getc
-	.globl	console_core_flush
+#if MULTI_CONSOLE_API
+	.globl console_pl011_register
 
+	/* -----------------------------------------------
+	 * int console_pl011_register(console_pl011_t *console,
+		uintptr_t base, uint32_t clk, uint32_t baud)
+	 * Function to initialize and register a new PL011
+	 * console. Storage passed in for the console struct
+	 * *must* be persistent (i.e. not from the stack).
+	 * In: x0 - pointer to empty console_pl011_t struct
+	 *     x1 - UART register base address
+	 *     w2 - UART clock in Hz
+	 *     w3 - Baud rate
+	 * Out: return 1 on success, 0 on error
+	 * Clobber list : x0, x1, x2, x6, x7, x14
+	 * -----------------------------------------------
+	 */
+func console_pl011_register
+	mov	x7, x30
+	cbz	x0, register_fail
+	str	x1, [x0, #CONSOLE_T_PL011_BASE]
+
+	/* shift registers around to match ABI of legacy init function */
+	mov	x6, x0
+	mov	x0, x1
+	mov	w1, w2
+	mov	w2, w3
+
+	bl	console_pl011_init
+	cbz	x0, register_fail
+
+	mov	x0, x6
+	adrp	x1, console_pl011_putc
+	add	x1, x1, :lo12:console_pl011_putc
+	str	x1, [x0, #CONSOLE_T_PUTC]
+	adrp	x1, console_pl011_getc
+	add	x1, x1, :lo12:console_pl011_getc
+	str	x1, [x0, #CONSOLE_T_GETC]
+	adrp	x1, console_pl011_flush
+	add	x1, x1, :lo12:console_pl011_flush
+	str	x1, [x0, #CONSOLE_T_FLUSH]
+	mov	x1, #(1 << CONSOLE_FLAG_BOOT_SHIFT | \
+		      1 << CONSOLE_FLAG_CRASH_SHIFT)
```

Comment:
We need the caller to be able to specify the console flags instead of hardcoding this here. Same for the other drivers

Reply:
Using the new API in the Arm platform case, we would want to:

* Register one serial console as the boot/crash console and a 2nd serial console as the runtime console.
* Override bl31_plat_runtime_setup() to change the crash console from the 1st serial console to the 2nd (as well as switching to the runtime state).

So I don't think the default setting should be a property of the driver. It would be less confusing from an API perspective if the default were the same for all drivers, or the setting always had to be provided by the client (so that there is no default).